core/src/test/java/org/elasticsearch/plugins/PluginsMetaDataUpgraderTests.java

```diff
@@ -0,0 +1,162 @@
+        assertNotNull(upgrade.custom(CustomMetaData1.TYPE));
+        assertThat(((TestCustomMetaData) upgrade.custom(CustomMetaData1.TYPE)).getData(), equalTo("modified_data1"));
+        for (IndexMetaData indexMetaData : upgrade) {
+            assertTrue(metaData.hasIndexMetaData(indexMetaData));
+        }
+    }
+
+    public void testIndexMetaDataUpgrade() throws Exception {
+        MetaData metaData = randomMetaData(1);
+        PluginsMetaDataUpgrader pluginsMetaDataUpgrader = new PluginsMetaDataUpgrader(Collections.singletonList(
+            metaData1 -> MetaData.builder(metaData1)
+                .put(randomIndexMetaData().index(metaData1.getConcreteAllIndices()[0]))
+                .build()));
+        MetaData upgrade = pluginsMetaDataUpgrader.upgrade(metaData);
+        assertTrue(upgrade != metaData);
+        for (IndexMetaData indexMetaData : upgrade) {
+            assertFalse(metaData.hasIndexMetaData(indexMetaData));
+        }
+    }
+
+    public void testMultipleMetaDataUpgrade() throws Exception {
+        MetaData metaData = randomMetaData(randomIntBetween(1, 5));
+        Function<MetaData, MetaData> upgrader1 = metaData1 -> MetaData.builder(metaData1)
+            .putCustom(CustomMetaData1.TYPE, new CustomMetaData1("modified_data1"))
+            .build();
+        Function<MetaData, MetaData> upgrader2 = metaData1 -> MetaData.builder(metaData1)
+            .putCustom(CustomMetaData2.TYPE, new CustomMetaData2("modified_data2"))
+            .build();
+        PluginsMetaDataUpgrader pluginsMetaDataUpgrader = new PluginsMetaDataUpgrader(Arrays.asList(upgrader1, upgrader2));
+        MetaData upgrade = pluginsMetaDataUpgrader.upgrade(metaData);
+        assertTrue(upgrade != metaData);
+        assertNotNull(upgrade.custom(CustomMetaData1.TYPE));
+        assertThat(((TestCustomMetaData) upgrade.custom(CustomMetaData1.TYPE)).getData(), equalTo("modified_data1"));
+        assertNotNull(upgrade.custom(CustomMetaData2.TYPE));
+        assertThat(((TestCustomMetaData) upgrade.custom(CustomMetaData2.TYPE)).getData(), equalTo("modified_data2"));
+    }
+
+    public void testNOOPMetaDataUpgrade() throws Exception {
+        MetaData metaData = randomMetaData(randomIntBetween(1, 5));
+        PluginsMetaDataUpgrader pluginsMetaDataUpgrader = new PluginsMetaDataUpgrader(Arrays.asList(Function.identity(),
+            Function.identity()));
+        MetaData upgrade = pluginsMetaDataUpgrader.upgrade(metaData);
+        assertTrue(upgrade == metaData);
+    }
+
+    public void testUpgradeFailure() throws Exception {
+        MetaData metaData = randomMetaData(randomIntBetween(1, 5));
+        PluginsMetaDataUpgrader pluginsMetaDataUpgrader = new PluginsMetaDataUpgrader(Collections.singletonList(
+            metaData1 -> {
+                throw new IllegalStateException("unsupported metaData");
+            }
+        ));
+        expectThrows(IllegalStateException.class, () -> pluginsMetaDataUpgrader.upgrade(metaData));
```

Comment:
Please check the actual exception message (so we know it is not some other ISE than what we expected).